discussion_of_helm_probe_and_resource.md

```diff
@@ -0,0 +1,35 @@
+Let's use this file for a discussion of how to set up the Helm probe and resource in common helm chart, e.g. tgi, tei, etc.
+
+Below is the proposal from @eero-t :
+
+[1] What's IMHO needed to fix things for probe timings:
+
+- Common components' value files include different probe timing sections for CPU and for accelerators
+
+- Their deployment templates select one based on .Values.accelDevice value (empty for CPU)
```

Comment:
I'll make a PR which selects better probe timings and custom metrics based on *.accelDevice values.

Reply:
Agree that we need to set different values for different device, especially for CPU some models startup are really slow.
What's the problem if we don't introduce the extra *.accelDevice, and use current variables .Values.livenessProbe/.Values.readinessProbe/.Values.startupProbe in each chart?

Reply:
> What's the problem if we don't introduce the extra *.accelDevice, and use current variables .Values.livenessProbe/.Values.readinessProbe/.Values.startupProbe in each chart?

**Intent of the probes**

The intent of startup and liveness probes is to restart deadlocked (or otherwise frozen) pods, on assumption that restart gets them working again.  I.e. startup and liveness probes make sense if there are deadlock/freezing problems with the service, and startup actually gets rid of the problem (eventually scheduler backs off from re-starting the failing instance, which at least gets rid of its resource usage).

Readiness probes are intended to avoid routing traffic to services that are temporarily down.

See: https://kubernetes.io/docs/tasks/configure-pod-container/configure-liveness-readiness-startup-probes/

**Issues with too short timings**

Issues that I have seen with current OPEA probe values, when multiple instances get scheduled on same Xeon:
* **Startup**: Scaled up instances never serve any queries, they get restarted because their startup does not finish in time.  Meaning that they just consume CPU on re-startups and slow down the already working inference service instances instead of service requests
*  **Liveness**: Stressed services that are working fine, just a bit slow, get restarted, which similarly to startup probes, just makes the situation worse
* **Readiness**: Probe failures keep all instances mostly in non-Ready state i.e. Service object does not route any queries to them, instead they get buffered to single overworked instance, which increases the latencies

Situation was noticeably improved by increasing the probe timings.

I haven't seem any deadlocks, so at least in my setup both startup & liveness probes are just harmful.  Readiness probes can be useful though, if they're fine-tuned to balance traffic from overworked (non-Ready) instances to ones with free capacity (in Ready state).

**Issues with too long timings**

If restart can fix the issue, and timings are too long, too long timings mean that fixing the issue is unnecessarily delayed.  If no traffic is routed to Pod due to non-Ready state, that's just potential performance decrease, not a functional issue.

However, if pod keeps in Ready state despite Liveness probe failing, it means more queries being routed to non-working pod and being lost when it is eventually restarted => Liveness probes timeouts should not be shorter than Readiness ones, but they could be longer.

Reply:
Partly this is a problem of current OPEA Helm charts missing suitable per-model resource requests / affinity / scheduler topology / NRI policies.  If each model + service arg + device combo would specify well enough their resource needs, and they were isolated enough from each other when running on CPU, there would be less need for increasing the default probe timings for CPU so much.